### PR TITLE
[#IOPID-4006] Locked Profile Outbound Adapter

### PR DESCRIPTION
**Pull request by salvatorelaiso**
This pull request introduces a new outbound adapter for managing locked profiles using Azure Table Storage, along with comprehensive tests and integration into the application. The main focus is on supporting profile lock status checks, robust error handling, and seamless configuration for both local (Azurite) and production environments.

##### Locked Profiles Table Adapter Implementation

* Added a new `LockedProfilesDataTableAdapter` in `locked-profiles-data-table.adapter.ts` that implements the `LockedProfilesPort` interface, providing `isLocked` and `healthcheck` methods for checking and monitoring locked profiles in Azure Table Storage. The adapter uses a zod schema for entity validation and robust error handling with `neverthrow`. [[1]](diffhunk://#diff-ae76726a660e3c07532602b022b9f35c5b4dc44083911f5d4694dc23b42ff382R1-R92) [[2]](diffhunk://#diff-56cad8065d6fc74f8e9f3c23faa8b00a46ce6349b6e0caa1c2e3f652a943c1ffR1-R7)

##### Testing

* Introduced a comprehensive integration test suite (`locked-profiles.adapter.integration.test.ts`) that provisions tables on Azurite, seeds test data, and verifies adapter behavior for various lock states and error scenarios. [[1]](diffhunk://#diff-373378f65ada047d9e32a7795fb35771d3e48d42f561f50a364e99325347946fR1-R193) [[2]](diffhunk://#diff-3dae255a32df59dc28a7d779b510734fa018e15c30138286fbe8edf8ad4fdb2aR1-R30) [[3]](diffhunk://#diff-16f91b0e773204c4aa8f7942cc6346e4852f086b90a6cb4a99159f50e64cf0efR24-R38)
* Added unit tests for the adapter's core logic, including defensive error handling and edge cases, in `locked-profiles-data-table.adapter.test.ts`.

##### Application Integration

* Integrated the locked profiles adapter into the main Fastify app, wiring it into the health check endpoint and supporting both production and development (Azurite) environments with dynamic credential and connection string handling. [[1]](diffhunk://#diff-04b3d8c570b95bf81412cc5cbd295b21dce8c95e5c24dcddf39fe8c8fdb0d7d4R1-R27) [[2]](diffhunk://#diff-04b3d8c570b95bf81412cc5cbd295b21dce8c95e5c24dcddf39fe8c8fdb0d7d4L18-R64)
* Updated the server config schema to include `NODE_ENV` for environment-specific logic.

##### Dependencies

* Added necessary Azure SDK and internal library dependencies to `package.json` to support Table Storage operations and schema validation.

These changes lay the foundation for managing locked profiles via Azure Table Storage, with a strong focus on testability, error safety, and environment-agnostic configuration.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a changeset, which I've added.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+import { LockedProfilesPort } from "../../domain/ports/outbound/locked-profiles.port.js";
+
+/**
+ * Schema for the LockedProfileDataTable entity.
+ * This schema is used to validate the structure of the data stored in the Azure Table Storage.
+ */
+export const LockedProfileDataTableSchema = z.object({
+  partitionKey: FiscalCodeSchema,
+  rowKey: z.string().regex(/^\d{9}$/, "rowKey must be 9 digits"),
```
Comment (arcogabbo) [inline apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: I think rowkey field can be transformed into a value-object

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import { ok, err, Result } from "neverthrow";
+import z from "zod";
+
+import { LockedProfilesPort } from "../../domain/ports/outbound/locked-profiles.port.js";
+
+/**
+ * Schema for the LockedProfileDataTable entity.
+ * This schema is used to validate the structure of the data stored in the Azure Table Storage.
+ */
+export const LockedProfileDataTableSchema = z.object({
```
Comment (arcogabbo) [inline apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: I think this structure can be moved to `entities` dir

File: packages/azure-sdk/src/index.ts
Comment (arcogabbo) [inline packages/azure-sdk/src/index.ts:1]: empty file

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+ *   userId: z.string(),
+ *   expiresAt: z.number().int().positive(),
+ * });
+ * ```
+ */
+export type TableEntitySchema = z.ZodObject<
+  {
+    partitionKey: z.ZodType<string>;
+    rowKey: z.ZodType<string>;
+  } & z.core.$ZodLooseShape
```
Comment (arcogabbo) [inline packages/azure-sdk/src/data-tables/data-table.wrapper.ts]: the LooseShape type makes PartitionKeyOf kind of types go unknown, forcing you to cast with `as` multiple times, defeating the purpose of this type helpers. I suggest to remove the LooseShape, in either case the `extends` keyword will make any following property valid
```suggestion
  }
```

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+    options?: GetTableEntityOptions,
+  ): Promise<Result<TableEntityWithMetadata<z.output<S>>, TableStorageError>> {
+    let response: TableEntityResult<Record<string, unknown>>;
+    try {
+      // pk/rk are string subtypes at the type level but TS's index-access
+      // through zod's shape widens to `unknown`; cast to satisfy the SDK.
+      response = await this.client.getEntity(
+        partitionKey as string,
+        rowKey as string,
+        options,
```
Comment (arcogabbo) [inline packages/azure-sdk/src/data-tables/data-table.wrapper.ts]: following previous comment, this benefits your types
```suggestion
      response = await this.client.getEntity(partitionKey, rowKey, options);

```

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    partitionKey: PartitionKeyOf<S>,
+    rowKey: RowKeyOf<S>,
+    options?: DeleteTableEntityOptions,
+  ): Promise<Result<DeleteTableEntityResponse, TableStorageError>> {
+    try {
+      const response = await this.client.deleteEntity(
+        partitionKey as string,
+        rowKey as string,
+        options,
+      );
```
Comment (arcogabbo) [inline packages/azure-sdk/src/data-tables/data-table.wrapper.ts]:
```suggestion
      const response = await this.client.deleteEntity(
        partitionKey,
        rowKey,
        options,
      );
```

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts (hunk 2 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: My take is that this schema is more of a persistence DTO than a domain entity: partitionKey/rowKey are Azure Table Storage concerns to me.
I think moving it to domain/entities/ would end up making the domain depend on the storage model.

I think the proper domain entity here would be a storage-agnostic `LockedProfile { fiscalCode, unlockCode }`, with the adapter mapping `LockedProfileDataTable` to`LockedProfile`.
You can see an example of what I have in mind in [locked-profiles.adapter.integration.test.ts](https://github.com/pagopa/io-auth-n-identity-domain/blob/ed0076000647e6a3c6a5e246543b1d26164a4996/apps/io-session-manager-oi/src/__integrations__/__tests__/locked-profiles.adapter.integration.test.ts#L52-L119).

I am adding these changes to emphasize the point: ed0076000647e6a3c6a5e246543b1d26164a4996

What do you think?

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts (hunk 3 above)
Comment (salvatorelaiso) [reply packages/azure-sdk/src/data-tables/data-table.wrapper.ts]: Thanks, resolved in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/861954f8fac3ab060d34953d6597077caab03320

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts (hunk 4 above)
Comment (salvatorelaiso) [reply packages/azure-sdk/src/data-tables/data-table.wrapper.ts]: Resolved in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/861954f8fac3ab060d34953d6597077caab03320

File: packages/azure-sdk/src/data-tables/data-table.wrapper.ts (hunk 5 above)
Comment (salvatorelaiso) [reply packages/azure-sdk/src/data-tables/data-table.wrapper.ts]: Resolved in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/861954f8fac3ab060d34953d6597077caab03320

File: packages/azure-sdk/src/index.ts
Comment (salvatorelaiso) [reply packages/azure-sdk/src/index.ts:1]: Removed in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/12e419c31ecce2dd1ae12566bf7b3653423820df

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts (hunk 1 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: Given https://github.com/pagopa/io-auth-n-identity-domain/pull/752#discussion_r3637604523, do you think it's acceptable this change? https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/6bea9bc3688e7aa4b44870856b20ff90c4617a2c

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        },
+      )) {
+        // If there's an error retrieving the entity, return the error
+        if (entity.isErr()) {
+          return err(entity.error);
+        }
+        // If we find at least one entity that is not released, the profile is locked
+        if (entity.value.entity.Released !== true) {
+          return ok(true);
+        }
```
Comment (gquadrati) [inline apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: The `Released !== true` check inside the loop is redundant: the filter `not Released` already guarantees that any returned entity has Released set to false or absent. The inner condition can never be true, so the check can be safely removed and replaced with a direct return ok(true).

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+// ---------------------------------------------------------------------------
+
+describe("LockedProfilesDataTableAdapter#isLocked", () => {
+  it("returns ok(false) when the iterator yields nothing (no locks found)", async () => {
+    listEntitiesMock.mockReturnValue(asyncIterableOf([]));
+
+    const result = await adapter.isLocked(FISCAL_CODE);
+
+    expect(result.isOk()).toBe(true);
+    expect(result._unsafeUnwrap()).toBe(false);
```
Comment (gquadrati) [inline apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: I think this is more readable, can we change everywhere?
```suggestion
    expect(result).toEqual(ok(false));
```

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+    expect(result.isOk()).toBe(true);
+    expect(result._unsafeUnwrap()).toBe(true);
+  });
+
+  it("skips entities where Released === true and returns ok(false) if none remain", async () => {
+    // Defensive: the OData filter already excludes released rows, but if a
+    // released row leaks through the adapter must not report it as a lock.
+    listEntitiesMock.mockReturnValue(
+      asyncIterableOf([okEntity({ Released: true })]),
```
Comment (gquadrati) [inline apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: This scenario cannot happen, since we got the filter on `not Released`

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    const result = await adapter.isLocked(FISCAL_CODE);
+
+    expect(result.isOk()).toBe(true);
+    expect(result._unsafeUnwrap()).toBe(false);
+  });
+
+  it("returns ok(true) when a non-released entity follows a released one", async () => {
+    listEntitiesMock.mockReturnValue(
+      asyncIterableOf([
+        okEntity({ Released: true, rowKey: "111111111" }),
```
Comment (gquadrati) [inline apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: ditto: This result cannot happen, since we got the filter on `not Released`

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    const notFound = new NotFoundError(TABLE_NAME, "table missing");
+    listEntitiesMock.mockReturnValue(asyncIterableOf([err(notFound)]));
+
+    const result = await adapter.isLocked(FISCAL_CODE);
+
+    expect(result.isErr()).toBe(true);
+    expect(result._unsafeUnwrapErr()).toBe(notFound);
+  });
+
+  it("stops iterating on the first err (does not inspect subsequent entities)", async () => {
```
Comment (gquadrati) [inline apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: interesting, can we do the same for an happy path, checking that it stops after the first matching element?

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts (hunk 6 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/locked-profiles-data-table.adapter.ts]: Done in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/0e193abd72e52084331dfd597dd1f514ee8e4c94

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 8 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Removed in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/6bfb14040d38d5904ce7fbaec13041b584fb597c

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 9 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Removed in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/dec1ff9b9fc10139b867da5efb984b808d8d2622

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 7 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Please check if this is good enough https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/c229cf683262d839434bdd84d5111f819825d10a

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 10 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Does this look good to you? https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/0b3f472cbc2546c927193937009cc47c3263b04c

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 7 above)
Comment (gquadrati) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: We can remove the line `expect(result.isOk()).toBe(true);`

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 10 above)
Comment (gquadrati) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: `Released: true` is never returned, just put two ` Released: false`

**Review by gquadrati — APPROVED** [review]
Just two small comments, but LGTM

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 7 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Removed in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/967e370f60575a5f2d65a8e36a5d160086d168a0

File: apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts (hunk 10 above)
Comment (salvatorelaiso) [reply apps/io-session-manager-oi/src/adapters/outbound/__tests__/locked-profiles-data-table.adapter.test.ts]: Done in https://github.com/pagopa/io-auth-n-identity-domain/pull/752/commits/98e4b63a5e25162f34314b4ab875e4adf96a2e80
